circuit_knitting_toolbox/circuit_cutting/cutting_evaluation.py

```diff
@@ -82,3 +183,185 @@ def append_measurement_circuit(
+        }
+
+    # Sample the joint quasiprobability decomposition
+    random_samples = generate_qpd_samples(bases, num_samples=num_samples)
+
+    # Calculate terms in coefficient calculation
+    kappa = np.prod([basis.kappa for basis in bases])
+    num_samples = sum([value[0] for value in random_samples.values()])
+
+    # Sort samples in descending order of frequency
+    sorted_samples = sorted(random_samples.items(), key=lambda x: x[1][0], reverse=True)
+
+    # Generate the outputs -- sub-experiments, coefficients, and frequencies
+    subexperiments: list[list[list[QuantumCircuit]]] = []
+    coefficients = []
+    sampled_frequencies = []
+    for z, (map_ids, sample_weight) in enumerate(sorted_samples):
+        redundancy = sample_weight[0]
+        subexperiments.append([])
+        actual_coeff = np.prod(
+            [basis.coeffs[map_id] for basis, map_id in strict_zip(bases, map_ids)]
+        )
+        sampled_coeff = (redundancy / num_samples) * (kappa * np.sign(actual_coeff))
+        coefficients.append(sampled_coeff)
+        sampled_frequencies.append(redundancy)
+        for i, (subcircuit, label) in enumerate(
+            strict_zip(subcircuit_list, sorted(subsystem_observables.keys()))
+        ):
+            map_ids_tmp = map_ids
+            if is_separated:
+                map_ids_tmp = tuple(map(map_ids.__getitem__, subcirc_map_ids[i]))
+            decomp_qc = decompose_qpd_instructions(
+                subcircuit, subcirc_qpd_gate_ids[i], map_ids_tmp
+            )
+            subexperiments[-1].append([])
+            so = subsystem_observables[label]
+            for j, cog in enumerate(so.groups):
+                meas_qc = append_measurement_circuit(decomp_qc, cog)
+                # Should have strictly 2 classical registers, "qpd measurements" and "observable_measurements"
+                subexperiments[-1][-1].append(meas_qc)
+
+    return subexperiments, coefficients, sampled_frequencies
+
+
+def _run_experiments_batch(
+    subexperiments: Sequence[Sequence[QuantumCircuit]],
+    backend: Backend,
+    shots: int,
+) -> list[list[tuple[dict[str, int], int]]]:
+    """Run subexperiments on the backend."""
+    counts: list[list[tuple[dict[str, int], int]]] = []
+    num_qpd_bits_flat = []
+
+    # Run all the experiments in one big batch
+    experiments_flat = list(chain.from_iterable(subexperiments))
+    for circ in experiments_flat:
+        if len(circ.cregs) < 1 or circ.cregs[-1].name != "observable_measurements":
+            raise ValueError(
+                "Circuit's last register is expected to be named 'observable_measurements'."
+            )
+        if len(circ.cregs) > 1 and circ.cregs[-2].name != "qpd_measurements":
+            raise ValueError(
+                "Circuit's next-to-last register is expected to be named 'qpd_measurements'."
+            )
+
+        if len(circ.cregs) < 2:
+            num_qpd_bits_flat.append(0)  # No gates decomposed to a measurement
```

Comment:
If we enforce that input circuits cannot contain cregs, I believe we can infer the number of qpd bits during reconstruction by just looking at the size of the corresponding observable. Under this restriction, the number of qpd bits should always be `total_clbits - len(sub-observable)`, right @garrison ?

Reply:
I played with this. I think there is something better we can do, but I'd rather think about it later. It'd be better if the observable measurements and clbits could be added before the midcircuit clbits. That would allow us to use the `cog` in `process_outcomes` to infer which clbits are qpd bits and drop the `num_qpd_bits` arg